libbindgen/src/codegen/mod.rs

```diff
@@ -516,6 +516,30 @@ impl CodeGenerator for Type {
                     inner_item.to_rust_ty(ctx)
                 };
 
+                {
+                    // FIXME(emilio): This is a workaround to avoid generating
+                    // incorrect type aliases because of types that we haven't
+                    // been able to resolve (because, eg, they depend on a
+                    // template parameter).
+                    //
+                    // It's kind of a shame not generating them even when they
+                    // could be referenced, but we already do the same for items
+                    // with invalid template parameters, and at least this way
+                    // they can be replaced, instead of generating plain invalid
+                    // code.
+                    let inner_canon_type =
+                        inner_item.expect_type().canonical_type(ctx);
+                    if inner_canon_type.is_named() &&
+                        inner_canon_type.name().map_or(false, |name| {
+                            name.contains("decltype ") ||
+                            name.contains("decltype(")
+                        }) {
```

Comment:
We should never have any spaces or parens for named types, regardless of decltype or typename.

What we want here is a generic check that this named type is a valid C++ identifier.